xpti/src/xpti_proxy.cpp

```diff
@@ -0,0 +1,353 @@
+            : m_loaded(false)
+            , m_fw_plugin_handle(nullptr)
+        {
+            // When this object is created, we attempt to load
+            // the share object implementation. We look for the
+            // environment variable XPTI_FRAMEWORK_DISPATCHER to 
+            // see if it has been set. If not, all methods in 
+            // the proxy should end up being close to no-ops
+            //
+            std::wstring env = m_loader.getEnvironmentVariable("XPTI_FRAMEWORK_DISPATCHER");
+            if (env.empty())
+                return;
+            std::string error;
+            m_fw_plugin_handle = m_loader.loadLibrary(&env[0], error);
+            if (m_fw_plugin_handle) {
+                // We will defer changing m_loaded = true until the
+                // end of this block after we are able to resolve 
+                // all of the entry points
+                //
+                m_dispatch_table.resize(XPTI_FW_API_COUNT);
+                for (auto &func_name : m_function_names) {
+                    xpti_plugin_function_t func = m_loader.findFunction(m_fw_plugin_handle, func_name.second);
+                    if (!func) { // Return if we fail on even one function 
+                        m_loader.unloadLibrary(m_fw_plugin_handle);
+                        m_fw_plugin_handle = nullptr;
+                        return;
+                    }
+                    m_dispatch_table[func_name.first] = func;
+                }
+                // Only if all the functions are found and loaded,
+                // do we set the m_loaded = true
+                //
+                m_loaded = true;
+            }
+        }
+
+        ~ProxyLoader() {
+            // If the loading of the framework library was 
+            // successful, we should close the handle in the
+            // destructor to decrement the reference count 
+            // maintained by the loader.
+            //
+            if (m_fw_plugin_handle) {
+                m_loader.unloadLibrary(m_fw_plugin_handle);
+            }
+        }
+
+        inline bool noErrors() {
+            return m_loaded;
+        }
+
+        void *functionByIndex(int index) {
+            if (index >= XPTI_INITIALIZE && index < XPTI_FW_API_COUNT) {
+                return m_dispatch_table[index];
+            }
+            return nullptr;
+        }
+    private:
+
+        bool                        m_loaded;
+        xpti_plugin_handle_t        m_fw_plugin_handle;
+        dispatch_table_t            m_dispatch_table;
+        xpti::utils::PlatformHelper m_loader;
+    };
+
+    static ProxyLoader g_loader;
+}
+
+XPTI_EXPORT_API uint16_t xptiRegisterUserDefinedTracePoint
+(
+    const char *tool_name,
+    uint8_t     user_defined_tp
+) {
+    if (xpti::g_loader.noErrors()) {
+        void *f = xpti::g_loader.functionByIndex(XPTI_REGISTER_USER_DEFINED_TP);
+        if (f) {
+            return (*(xpti_register_user_defined_tp_t)f)

```

Comment:
You can simplify a lot all the code below.
I think it would be nice to have a templated interface so you can just write above
```
auto f = xpti::g_loader.functionByIndex<xpti_register_user_defined_tp_t>(XPTI_REGISTER_USER_DEFINED_TP);
```
or even
```
auto f = xpti::g_loader.functionByIndex<XPTI_REGISTER_USER_DEFINED_TP>();
```
by doing some metaprogramming magic.
It is very common in the SYCL API, for example using the "4.4 Param traits class"

At the end the full function could be even just
```
return xpti::g_loader.exec_or_error<XPTI_REGISTER_USER_DEFINED_TP>(tool_name, user_defined_tp);
```
:-)

Reply:
I like the suggestion - I would love to visit this in the next round of changes after the merge.